lib/galaxy/jobs/__init__.py

```diff
@@ -1307,6 +1333,7 @@ def change_state(self, state, info=False, flush=True, job=None):
             return
         for dataset_assoc in job.output_datasets + job.output_library_datasets:
             dataset = dataset_assoc.dataset
+            model.StorageMedia.refresh_all_media_credentials(dataset.dataset.active_storage_media_associations, self.app.authnz_manager, self.sa_session)
```

Comment:
This piece of code has also been heavily optimized - I'm really quite nervous about this extra association fetch per output. For huge collection outputs - this could result in a huge performance hit I would guess.

I'm not immediately sure how to handle this though. I already feel like this refresh around every operation is problematic - it seems like it should be push a layer down so these function calls don't have to worry about these details - but I'm not sure how to appropriately push the authnz_manager down into that level.

Reply:
@jmchilton would https://github.com/VJalili/galaxy/commit/ca7e71609b5b08b7ee870e0452dd663b4e399372 be in the direction to address your performance hit concerns?